src/__tests__/__snapshots__/preval.test.ts.snap

```
@@ -765,7 +765,20 @@ Dependencies: NA
 `;
 
 exports[`extractor non-hoistable identifiers 1`] = `
-"<<DIRNAME>>/source.js: An error occurred when evaluating the expression: days is not defined. Make sure you are not using a browser or Node specific API.
+"<<DIRNAME>>/source.js: An error occurred when evaluating the expression: 
+
+  > days is not defined. 
+
+  Make sure you are not using a browser or Node specific API and all the variables are available in static context.
+  Linaria have to extract pieces of your code to resolve the interpolated values.
+  Defining styled component or class will not work inside:
```

Instead of showing on big error here, we can easily modify `extract.ts` in order to check that `TaggedTemplateExpression` is placed on the top level of a file.

---

It makes some sense, but the evaluation will work if the user will not use any dependencies. So we should do it only for lazy dependencies, do I understand correctly?

---

And it will work in some cases like this
```js
const a = () => {
  const color = 'red'
  return styled.h1`
    color: ${color}
  `
}
```

---

It will work but in an unpredictable way. For example, what if a referenced variable is defined in both scopes?

---

Good point. Do you mean to remove the possibility to use Linaria outside the top-level scope of the file? 
That would be a quite huge breaking change but in terms of stability, it would improve much. I think we should discuss it more. 
What is your idea to handle that, if we check that `TaggedTemplateExpression` is out of top-level scope, what then? until it does not raise an error we shouldn't do anything to not introduce breaking change.

---

It's the same situation as with hooks in React: is it possible to use it in loops? Yes, it is. Is it safe? No.
There is only one safe case when you can use Linaria out of top-level scope: all interpolated variables are defined in a top-level scope. But that means that you can easily move styled/css call to the top-level. So I would prefer to throw an error if out of top-level case is detected.
For usage inside components, we can probably introduce some `useStyles` hook and, if I properly remember, it was even discussed here.

---

Throwing an error is a good idea, but Let's first discuss it internally. We can merge this error reporting to both `2.x` and `1.4.x` and then If we decide to restrict usage to top-level scope I would like to work on this :D

---

Ok, let's merge it! :)